lib/rust/ensogl/core/src/application/frp.rs

```diff
@@ -506,3 +506,536 @@ macro_rules! define_endpoints_emit_alias {
+
+        Input { $([$($input_opts:tt)*])?
+            $($(#[doc=$($in_doc :tt)*])*
+            $in_field : ident ($($in_field_type : tt)*)),* $(,)?
+        }
+
+        Output { $([$($output_opts:tt)*])?
+            $($(#[doc=$($out_doc:tt)*])*
+            $out_field : ident ($($out_field_type : tt)*)),* $(,)?
+        }
+    ) => {
+        use $crate::frp::IntoParam;
+
+        /// Frp network and endpoints.
+        #[derive(Debug,Derivative)]
+        #[derive(CloneRef, Clone)]
+        #[allow(missing_docs)]
+        pub struct Frp $(<$($param $(:$($constraints)*)?),*>)? {
+            public: api::Public $(<$($param),*>)?,// deref
+            private: Rc<api::Private $(<$($param),*>)?>,
+        }
+
+         impl $(<$($param $(:$($constraints)*)?),*>)? Frp $(<$($param),*>)? {
+            /// Create Frp endpoints within and the associated network.
+            pub fn new() -> Self {
+                let network = $crate::frp::Network::new(file!());
+                let public_input = api::public::Input::new(&network);
+                let private_input = api::private::Input::new(&network, &public_input);
+                let private_output = api::private::Output::new(&network);
+                let public_output = api::public::Output::new(&network, &private_output, &public_input);
+                let combined = api::public::Combined::new(&public_input,&public_output);
+                let public  = api::Public{ input: public_input, output: public_output, combined };
+                let private  = Rc::new(api::Private{ network, input: private_input, output: private_output});
+                Self {public,private}
+            }
+        }
+
+        impl $(<$($param $(:$($constraints)*)?),*>)? Default for Frp $(<$($param),*>)? {
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+
+          impl $(<$($param $(:$($constraints)*)?),*>)?  $crate::application::command::CommandApi for Frp $(<$($param),*>)?  {
+            fn command_api(&self)
+            -> Rc<RefCell<HashMap<String,$crate::application::command::Command>>> {
+                self.public.output.command_map.clone()
+            }
+
+            fn status_api(&self) -> Rc<RefCell<HashMap<String,$crate::frp::Sampler<bool>>>> {
+                self.public.output.status_map.clone()
+            }
+        }
+
+        impl $(<$($param $(:$($constraints)*)?),*>)?  $crate::application::command::FrpNetworkProvider for Frp $(<$($param),*>)?  {
+            fn network(&self) -> &$crate::frp::Network { &self.private.network }
+        }
+
+        impl $(<$($param $(:$($constraints)*)?),*>)? Deref for Frp $(<$($param),*>)? {
+                    type Target = api::Public $(<$($param),*>)?;
+            fn deref(&self) -> &Self::Target {
+                &self.public
+            }
+        }
+
+        #[allow(missing_docs)]
+        #[allow(clippy::manual_non_exhaustive)]
+        pub mod api {
```

Comment:
docs really need to explain that. I remember we've been talking about some modules, but although I was involved in designing that, I have trouble understanding what is happening here now. Proper docs would be a life-saver.

Reply:
Added more docs here. Let me know whether they are helpful or missing some information that you would be looking for here.